[schedulers] hanlde dtype in add_noise

patil-suraj opened the pull request:
Make sure that the `alphas/sigmas`  have the same `dtype` as the input samples.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_767). All of your documentation changes will be reflected on that endpoint.

kashif commented:
@patil-suraj should i add tests for fp16 schedulers?

NouamaneTazi commented on src/diffusers/models/vae.py:
```diff
@@ -337,12 +337,16 @@ def __init__(self, parameters, deterministic=False):
         self.var = torch.exp(self.logvar)
         if self.deterministic:
-            self.var = self.std = torch.zeros_like(self.mean).to(device=self.parameters.device)
+            self.var = self.std = torch.zeros_like(self.mean).to(
+                device=self.parameters.device, dtype=self.parameters.dtype
+            )
```
Better use
```
torch.zeros_like(self.mean).to(
                device=self.parameters.device, dtype=self.parameters.dtype
            )
```
to create tensor directly on device with correct dtype

NouamaneTazi commented on src/diffusers/models/vae.py:
```diff
@@ -337,12 +337,16 @@ def __init__(self, parameters, deterministic=False):
         device = self.parameters.device
         sample_device = "cpu" if device.type == "mps" else device
-        sample = torch.randn(self.mean.shape, generator=generator, device=sample_device).to(device)
+        sample = torch.randn(self.mean.shape, generator=generator, device=sample_device).to(
+            device=device, dtype=self.parameters.dtype
+        )
```
Same as previous comment

kashif replied:
no here it's fine for gpu/cpu reproducability

NouamaneTazi replied:
Sorry im not aware of this. Does having the following create reproducability issues? Can you point me to a related PR/issue if it's the case?
```python
torch.randn(self.mean.shape, generator=generator, device=sample_device, dtype=self.parameters.dtype)
```

kashif replied:
I believe when the generator is not None, doing randn on the GPU with the same generator will give different results than on the CPU and so we currently make the noise etc. on the CPU with the generator and then move it to the device...  see for example https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_ddim.py#L287

NouamaneTazi replied:
Thanks for the explanation 🙏🏼

NouamaneTazi reviewed (approved):
LGTM :)

NouamaneTazi commented on src/diffusers/schedulers/scheduling_ddim.py:
```diff
@@ -306,6 +306,10 @@ def add_noise(
             timesteps = timesteps.to(original_samples.device)
 
+        # Make sure that alphas_cumprod are in the same dtype as the samples
+        if self.alphas_cumprod.dtype != original_samples.dtype:
+            self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
```
Is the `if` necessary here?

NouamaneTazi commented on src/diffusers/schedulers/scheduling_ddpm.py:
```diff
@@ -300,6 +300,10 @@ def add_noise(
             timesteps = timesteps.to(original_samples.device)
 
+        # Make sure that alphas_cumprod are in the same dtype as the samples
+        if self.alphas_cumprod.dtype != original_samples.dtype:
```
Is the `if` necessary here?

NouamaneTazi reviewed (commented):
I believe that
```python
        if self.alphas_cumprod.dtype != original_samples.dtype:
            self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
```
is equivalent to just doing
```python
self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
```
Same for `dtype`. And that could cut off the lines of code quite a bit

patil-suraj commented:
Removed the if as suggested by @NouamaneTazi ! The `vae` and `pipeline` shouldn't have been modified in the PR, removed those from here and will address suggestions related to that in #769

pcuenca commented:
> I believe that
> 
> ```python
>         if self.alphas_cumprod.dtype != original_samples.dtype:
>             self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
> ```
> 
> is equivalent to just doing
> 
> ```python
> self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
> ```
> 
> Same for `device`. And that could cut off the lines of code quite a bit

According to this comment it may have a penalty https://github.com/huggingface/diffusers/pull/511/files#r987896215. Is there any way we can now for sure?
